src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/convenience/key_vault_credential.py

```diff
@@ -7,18 +7,53 @@
 
 import adal
 import urllib
+from requests.auth import AuthBase
+import requests
 
 from msrest.authentication import Authentication
 
 from . import HttpBearerChallenge
 from . import HttpBearerChallengeCache as ChallengeCache
 
-class KeyVaultCredential(Authentication):
+class KeyVaultAuthBase(AuthBase):
 
-    def __init__(self, token):
-        self._token = token
+    def __init__(self, token_callback):
+        self._callback = token_callback
+        self._token = None
 
+    def __call__(self, request):
+        # attempt to pre-fetch token if challenge is cached
+        if self._callback:
+            challenge = ChallengeCache.get_challenge_for_url(request.url)
+            if challenge:
+                # if challenge cached, retrieve token and update the request
+                self.set_auth_header(request, challenge)
+            else:
+                # send the request to retrieve the challenge, then request the token and update
+                # the request
+                response = requests.Session().send(request)
```

Comment:
This is not critical, but would be really good if we can wire up the commons flag, say disable the cert validation for fiddler, and logging. Anyway we can get it from msrest pipeline?  //cc:@annatisch